src/include/detail/graph/vamana.h

```diff
@@ -0,0 +1,1019 @@
+   * the group?
+   */
+  auto write_index(const std::string& group_uri, bool overwrite = false) {
+    // copilot ftw!
+    // metadata: dimension, ntotal, L, R, B, alpha_min, alpha_max, medioid
+    // Save as a group: metadata, feature_vectors, graph edges, offsets
+
+    tiledb::Context ctx;
+    tiledb::VFS vfs(ctx);
+    if (vfs.is_dir(group_uri)) {
+      if (overwrite == false) {
+        return false;
+      }
+      vfs.remove_dir(group_uri);
+    }
+
+    tiledb::Config cfg;
+    tiledb::Group::create(ctx, group_uri);
+    auto write_group = tiledb::Group(ctx, group_uri, TILEDB_WRITE, cfg);
+
+    for (auto&& [name, value, type] : metadata) {
+      write_group.put_metadata(name, type, 1, value);
+    }
+
+    // feature_vectors
+    auto feature_vectors_uri = group_uri + "/feature_vectors";
+    write_matrix(ctx, feature_vectors_, feature_vectors_uri);
+    write_group.add_member("feature_vectors", true, "feature_vectors");
+
+    // adj_list
+    auto adj_scores_uri = group_uri + "/adj_scores";
+    auto adj_ids_uri = group_uri + "/adj_ids";
+    auto adj_index_uri = group_uri + "/adj_index";
+    auto adj_scores = Vector<score_type>(graph_.num_edges());
+    auto adj_ids = Vector<id_type>(graph_.num_edges());
+    auto adj_index = Vector<index_type>(graph_.num_vertices() + 1);
+
+    size_t edge_offset{0};
+    for (size_t i = 0; i < num_vertices(graph_); ++i) {
+      adj_index[i] = edge_offset;
+      for (auto&& [score, id] : graph_.out_edges(i)) {
+        adj_scores[edge_offset] = score;
+        adj_ids[edge_offset] = id;
+        ++edge_offset;
+      }
+    }
+    adj_index.back() = edge_offset;
+
+    write_vector(ctx, adj_scores, adj_scores_uri);
+    write_group.add_member("adj_scores", true, "adj_scores");
+
+    write_vector(ctx, adj_ids, adj_ids_uri);
+    write_group.add_member("adj_ids", true, "adj_ids");
+
+    write_vector(ctx, adj_index, adj_index_uri);
+    write_group.add_member("adj_index", true, "adj_index");
+
+    write_group.close();
+    return true;
+  }
+
+  /**
+   * @brief Log statistics about the index
+   */
+  void log_index() {
+#if 0
+    _count_data.insert_entry("dimension", dimension_);
+    _count_data.insert_entry("num_vectors", num_vectors_);
+    _count_data.insert_entry("L_build", L_build_);
+    _count_data.insert_entry("R_max_degree", R_max_degree_);
+    _count_data.insert_entry("num_edges", graph_.num_edges());
+    _count_data.insert_entry("num_comps", num_comps());
+    _count_data.insert_entry("num_visited_vertices", num_visited_vertices());
+
+    auto&& [min_degree, max_degree] =
+        minmax_element(begin(graph_), end(graph_), [](auto&& a, auto&& b) {
+          return a.size() < b.size();
+        });
+    _count_data.insert_entry("min_degree", min_degree->size());
+    _count_data.insert_entry("max_degree", max_degree->size());
+    _count_data.insert_entry(
+        "avg_degree",
+        (double)graph_.num_edges() / (double)num_vertices(graph_));
+#endif
```

Comment:
Why the ifdef? Can't we just not call this code but leave it in so that we don't accidentally break it?